tests/components/abode/test_alarm_control_panel.py

```diff
@@ -0,0 +1,84 @@
+"""Tests for the Abode alarm control panel device."""
+from unittest.mock import patch
+
+from homeassistant.components.abode import ATTR_DEVICE_ID
+from homeassistant.components.alarm_control_panel import DOMAIN as ALARM_DOMAIN
+from homeassistant.const import (
+    ATTR_ENTITY_ID,
+    ATTR_FRIENDLY_NAME,
+    ATTR_SUPPORTED_FEATURES,
+    SERVICE_ALARM_ARM_AWAY,
+    SERVICE_ALARM_ARM_HOME,
+    SERVICE_ALARM_DISARM,
+    STATE_ALARM_DISARMED,
+)
+
+from .common import setup_platform
+
+
+async def test_entity_registry(hass):
+    """Tests that the devices are registered in the entity registry."""
+    await setup_platform(hass, ALARM_DOMAIN)
+    entity_registry = await hass.helpers.entity_registry.async_get_registry()
+
+    entry = entity_registry.async_get("alarm_control_panel.abode_alarm")
+    # Abode alarm device unique_id is the MAC address
+    assert entry.unique_id == "001122334455"
+
+
+async def test_automation_attributes(hass):
+    """Test the alarm control panel attributes are correct."""
+    await setup_platform(hass, ALARM_DOMAIN)
+
+    state = hass.states.get("alarm_control_panel.abode_alarm")
+    assert state.state == STATE_ALARM_DISARMED
+    assert state.attributes.get(ATTR_DEVICE_ID) == "area_1"
+    assert not state.attributes.get("battery_backup")
+    assert not state.attributes.get("cellular_backup")
+    assert state.attributes.get(ATTR_FRIENDLY_NAME) == "Abode Alarm"
+    assert state.attributes.get(ATTR_SUPPORTED_FEATURES) == 3
+
+
+async def test_set_alarm_away(hass):
+    """Test the alarm control panel can be set to away."""
+    await setup_platform(hass, ALARM_DOMAIN)
+
+    with patch("abodepy.ALARM.AbodeAlarm.set_away") as mock_set_away:
+        await hass.services.async_call(
+            "alarm_control_panel",
+            SERVICE_ALARM_ARM_AWAY,
```

Comment:
I could get full coverage of alarm_control_panel.py in the Abode integration if I could figure out how to force HA to refresh the device state in tests. After this service is called, abodepy automatically forces an update to the device JSON so if I could just essentially call `schedule_update_ha_state()` for a specific device, I could then just check the state which would cover lines 41-46 in alarm_control_panel.py for test coverage.

Reply:
Patch `abode.events.add_device_callback` in the correct place so we get a handle on the callback that we pass in. Then we can call that callback to fake a device feedback.

Reply:
I think I know what I need to do, just can't seem to figure out how. The callback I want to call to force the device update in HA is `self._update_callback` in:
```
    async def async_added_to_hass(self):
        """Subscribe to device events."""
        self.hass.async_add_job(
            self._data.abode.events.add_device_callback,
            self._device.device_id,
            self._update_callback,
        )
        self.hass.data[DOMAIN].entity_ids.add(self.entity_id)
```
I believe I need to some how "capture" that callback with patching and manually call it.

This might be a bit over my head for now. I'll keep looking into it but not sure if it's worth holding up the PR for it. It may be a while 😄

Reply:
We may need to await hass.async_block_till_done if the callback schedules updates.

Reply:
Added `await hass.async_block_till_done()` but still the same thing, it doesn't appear to be updating the device state.

```
        mock_callback.call_args[0][1]("area_1")
        await hass.async_block_till_done()
        state = hass.states.get("alarm_control_panel.abode_alarm")
        assert state.state == "armed_away"
```

Alarm state is still showing as the original "disarmed" state.

Reply:
The callback should be scheduled on the executor thread pool since it's a sync funcion.
```py
update_callback = mock_callback.call_args[0][1]
await hass.async_add_executor_job(update_callback, "area_1")
await hass.async_block_till_done()
...
```

Reply:
That makes sense! Unfortunately, still no luck as the device state appears to not be updated.
```
        update_callback = mock_callback.call_args[0][1]
        await hass.async_add_executor_job(update_callback, "area_1")
        await hass.async_block_till_done()
        state = hass.states.get("alarm_control_panel.abode_alarm")
        assert state.state == "armed_away"
```

Reply:
Okay, so the issue is I'm patching `abodepy.ALARM.AbodeAlarm.set_away`, which obviously doesn't allow that actual method to execute in abodepy which is what changes the alarm state to away. The solution I thought up is this:

```
async def test_set_alarm_away(hass, requests_mock):
    """Test the alarm control panel can be set to away."""
    requests_mock.put(
        CONST.get_panel_mode_url("1", CONST.MODE_AWAY),
        json={"area": "1", "mode": CONST.MODE_AWAY},
    )
    with patch("abodepy.AbodeEventController.add_device_callback") as mock_callback:
        await setup_platform(hass, ALARM_DOMAIN)

        await hass.services.async_call(
            "alarm_control_panel",
            SERVICE_ALARM_ARM_AWAY,
            {ATTR_ENTITY_ID: "alarm_control_panel.abode_alarm"},
            blocking=True,
        )
        await hass.async_block_till_done()
        # Manually call the callback
        update_callback = mock_callback.call_args[0][1]
        await hass.async_add_executor_job(update_callback, "area_1")
        await hass.async_block_till_done()
        state = hass.states.get("alarm_control_panel.abode_alarm")
        assert state.state == "armed_away"
```

I'm no longer patching `abodepy.ALARM.AbodeAlarm.set_away` and using `requests_mock` to return a mock response. This way the `set_away` method is actually executed. The obvious difference is this is testing "deeper" into abodepy.

Reply:
I suggest we instead patch the `mode` property to return the expected correct mode on the device instance and schedule the callback.

Reply:
I'm not quite following. Are you saying to patch the [mode property in abodepy](https://github.com/MisterWil/abodepy/blob/c800d0a4dfc8a60a4513146ebbec2fc9a2ac7409/abodepy/devices/alarm.py#L113)? So in this case, I would be patching three things:

```
    with patch("abodepy.AbodeEventController.add_device_callback") as mock_callback:
        with patch("abodepy.ALARM.AbodeAlarm.set_away") as mock_set_away:
            with patch("abodepy.ALARM.AbodeAlarm.mode", side_effect="armed_away"):
```

And passing a side_effect with the mode so it returns the desired away status. When I try this, the device status ends up being unknown but I also am not following why this is better or preferred (I'm sure I'm just misunderstanding though). The [`set_mode` method in abodepy](https://github.com/MisterWil/abodepy/blob/c800d0a4dfc8a60a4513146ebbec2fc9a2ac7409/abodepy/devices/alarm.py#L34) is what is actually changing the device state within abodepy based on the JSON response object from the HTTP put request.

Reply:
Here's where we access the `is_away` property in the alarm_control_panel.

https://github.com/home-assistant/core/blob/7737387efe8f592a913fe8c39a6991f9266a0b78/homeassistant/components/abode/alarm_control_panel.py#L36-L47

And here's that property and the mode property in the alarm abode device:

https://github.com/MisterWil/abodepy/blob/c800d0a4dfc8a60a4513146ebbec2fc9a2ac7409/abodepy/devices/alarm.py#L107-L117

The `mode` property is a public interface on the abode alarm device in contrast to the request interface. We don't want to take responsibility for the api interface even in tests. The abode library may change the internal logic without notice but changing a public interface is hopefully communicated well.

Reply:
Makes sense! I completely missed that `mode` was being used in `is_away`...

The issue I'm having now is returning the appropriate `return_value`. I believe I have to use `PropertyMock`:

```
with patch("abodepy.ALARM.AbodeAlarm.mode", new_callable=PropertyMock) as mock_mode:
            mock_mode.return_value = "armed_away"

            update_callback = mock_callback.call_args[0][1]
            await hass.async_add_executor_job(update_callback, "area_1")
            await hass.async_block_till_done()
            mock_mode.assert_called()

            state = hass.states.get(DEVICE_ID)
            assert state.state == STATE_ALARM_ARMED_AWAY
```

But the test fails with `AssertionError: assert 'unknown' == 'armed_away'`. If I try to mock `is_away`, the test fails with `AssertionError: assert 'disarmed' == 'armed_away'`.